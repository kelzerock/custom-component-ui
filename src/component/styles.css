html,
body {
  height: 100%;
  margin: 0;
}

.wrapper {
  margin: 0 auto;
  max-width: 800px;
}
