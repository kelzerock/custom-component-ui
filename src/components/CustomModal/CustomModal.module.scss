.wrapper {
  padding: 2px;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #00000040;
  backdrop-filter: blur(1px);
}

.modal {
  margin: 20px 10px;
  padding: 10px;
  background-color: var(--color-white);
  border-radius: 5px;
  max-width: 80%;
  max-height: 80%;
  overflow-y: auto;
}

.lock {
  overflow: hidden;
}