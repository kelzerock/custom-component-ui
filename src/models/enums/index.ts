export { Checkbox } from "./enums-for-test/Checkbox";
