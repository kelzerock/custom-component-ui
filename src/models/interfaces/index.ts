export { CustomSelectProps } from "./custom-select-props";
export { CustomButtonProps } from "./custom-button-props";
export { CustomTextFieldProps } from "./custom-text-field-props";
export { OptionsData } from "./option-data";
