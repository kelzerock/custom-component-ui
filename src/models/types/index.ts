export { ButtonSize } from "./button-size";
export { ButtonVariant } from "./button-variant";
