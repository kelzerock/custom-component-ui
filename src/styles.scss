:root {
  --color-main-blue: #1976d2;
  --color-main-blue-100: #2070c0;
  --color-white: #fff;
  --color-white-100: #f6fafd;
  --color-red: #d32f2f;
  --color-gray: #f0f0f0;
  --color-dark-gray: #00000099;
}