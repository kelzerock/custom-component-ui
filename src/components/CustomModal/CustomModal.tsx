import { FC, useEffect } from "react";
import { CustomModalProps } from "../../models/interfaces/custom-modal-props";
import styles from "./CustomModal.module.scss";
import { Modal } from "../../models/enums";
import clsx from "clsx";

export const CustomModal: FC<CustomModalProps> = ({
  onClose,
  open,
  children,
  onClick,
  className,
  ...props
}: CustomModalProps) => {
  useEffect(() => {
    if (open) document.body.classList.add(styles.lock);

    return () => {
      document.body.classList.remove(styles.lock);
    };
  }, [open]);

  const handleClick = (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    if (event.target === event.currentTarget) onClose();
  };

  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === "Escape") {
        onClose();
      }
    };

    document.body.addEventListener("keydown", handleKeyDown);

    return () => {
      document.body.removeEventListener("keydown", handleKeyDown);
    };
  }, [onClose]);

  if (!open) return null;

  return (
    <div
      className={styles.wrapper}
      onClick={onClick ?? handleClick}
      data-testid={Modal.wrapper}
    >
      <div
        className={clsx(styles.modal, className)}
        {...props}
        data-testid={Modal.modal}
      >
        {children}
      </div>
    </div>
  );
};
