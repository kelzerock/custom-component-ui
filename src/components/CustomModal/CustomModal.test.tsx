import { render, screen } from "@testing-library/react";
import { useState } from "react";
import { CustomModal } from "./CustomModal";
import userEvent from "@testing-library/user-event";
import { Modal } from "../../models/enums";
import styles from "./CustomModal.module.scss";

const modalContent = "Modal content";
const TestComponent = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <>
      <button onClick={() => setIsOpen(true)} data-testid={"button"}>
        Open modal
      </button>
      <CustomModal open={isOpen} onClose={() => setIsOpen(false)}>
        {modalContent}
      </CustomModal>
    </>
  );
};

describe("Check CustomModal component", () => {
  it("Should open the modal when clicking the button", async () => {
    render(<TestComponent />);

    const button = screen.getByTestId("button");
    await userEvent.click(button);

    expect(screen.getByText(modalContent)).toBeInTheDocument();
  });

  it("should close the modal when clicking outside", async () => {
    render(<TestComponent />);

    const button = screen.getByTestId("button");
    await userEvent.click(button);

    const wrapper = screen.getByTestId(Modal.wrapper);
    await userEvent.click(wrapper);

    expect(screen.queryByText(modalContent)).not.toBeInTheDocument();
  });

  it("should block scrolling when modal is open", async () => {
    render(<TestComponent />);

    const button = screen.getByText("Open modal");
    await userEvent.click(button);

    expect(document.body).toHaveClass(styles.lock);
  });

  it("should remove scroll block when modal is closed", async () => {
    render(<TestComponent />);

    const button = screen.getByText("Open modal");
    await userEvent.click(button);

    const wrapper = screen.getByTestId(Modal.wrapper);
    await userEvent.click(wrapper);

    expect(document.body).not.toHaveClass(styles.lock);
  });

  it("should close when Escape key is pressed", async () => {
    render(<TestComponent />);

    const button = screen.getByText("Open modal");
    await userEvent.click(button);

    await userEvent.keyboard("{Escape}");

    expect(screen.queryByText(modalContent)).not.toBeInTheDocument();
  });

  test("check UI snapshot", async () => {
    const { container } = render(<TestComponent />);
    const button = screen.getByText("Open modal");
    await userEvent.click(button);
    expect(container).toMatchSnapshot();
  });
});
