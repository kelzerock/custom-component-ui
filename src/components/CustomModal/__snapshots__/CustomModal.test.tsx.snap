// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check CustomModal component check UI snapshot 1`] = `
<div>
  <button
    data-testid="button"
  >
    Open modal
  </button>
  <div
    class="wrapper"
    data-testid="wrapper"
  >
    <div
      class="modal"
      data-testid="modal"
    >
      Modal content
    </div>
  </div>
</div>
`;
